chalice/utils.py

```diff
@@ -96,6 +97,53 @@ def serialize_to_json(data):
     return json.dumps(data, indent=2, separators=(',', ': ')) + '\n'
 
 
+class ChaliceZipFile(zipfile.ZipFile):
+    """Support deterministic zipfile generation.
+
+    Normalizes datetime and permissions.
+
+    """
+
+    compression = 0  # Try to make mypy happy.
+    _default_time_time = (1980, 1, 1, 0, 0, 0)
+
+    def __init__(self, *args, **kwargs):
+        # type: (Any, Any) -> None
+        self._osutils = cast(OSUtils, kwargs.pop('osutils', OSUtils()))
+        super(ChaliceZipFile, self).__init__(*args, **kwargs)
+
+    # pylint: disable=W0221
+    def write(self, filename, arcname=None, compress_type=None):
+        # type: (Text, Optional[Text], Optional[int]) -> None
+        # Only supports files, py2.7 and 3 have different signatures.
+        # We know that in our packager code we never call write() on
+        # directories.
+        zinfo = self._create_zipinfo(filename, arcname, compress_type)
+        with open(filename, 'rb') as f:
+            self.writestr(zinfo, f.read())
+
+    def _create_zipinfo(self, filename, arcname, compress_type):
+        # type: (Text, Optional[Text], Optional[int]) -> zipfile.ZipInfo
+        # The main thing that prevents deterministic zip file generation
+        # is that the mtime of the file is included in the zip metadata.
+        # We don't actually care what the mtime is when we run on lambda,
+        # so we always set it to the default value (which comes from
+        # zipfile.py).  This ensures that as long as the file contents don't
+        # change (or the permissions) then we'll always generate the exact
+        # same zip file bytes.
+        # We also can't use ZipInfo.from_file(), it's only in python3.
+        st = self._osutils.stat(str(filename))
+        if arcname is None:
+            arcname = filename
+        arcname = self._osutils.normalized_filename(str(arcname))
+        arcname = arcname.lstrip(os.sep)
+        zinfo = zipfile.ZipInfo(arcname, self._default_time_time)
+        zinfo.external_attr = (st.st_mode & 0xFFFF) << 16
```

Comment:
can we get a comment in here what this is doing?  or can we use the octal form here (assuming 2.7 is dropped).

originally this was set to ensure files were always readable in lambda, but then the question of executable bins in uploads came up, part of the original was i've come across many enterprise environments where per some cis or stig benchmark they do default restrictive umask, that breaks lambda if just respecting file perms, it also has the bonus of not considering them for file delta. manual construction of the octal with a check of 747 or 644 on the basis of extant file executable would read more cleanly here.

Reply:
confirmed that behavior is still current

```python
import os
from pwd import getpwuid

def lambda_handler(event, context):
    stat = os.stat(__file__)
    print("File Owner %r" % ( getpwuid(stat.st_uid),))
    print("Executor %r" % ( getpwuid(os.geteuid()),))

    return 22
```

```
START RequestId: 4d17de89-800b-4871-8d17-063b5c233aa6 Version: $LATEST
File Owner pwd.struct_passwd(pw_name='slicer', pw_passwd='x', pw_uid=498, pw_gid=497, pw_gecos='', pw_dir='/tmp', pw_shell='/sbin/nologin')
Executor pwd.struct_passwd(pw_name='sbx_user1051', pw_passwd='x', pw_uid=496, pw_gid=495, pw_gecos='', pw_dir='/home/sbx_user1051', pw_shell='/sbin/nologin')
END RequestId: 4d17de89-800b-4871-8d17-063b5c233aa6
```

Reply:
>or can we use the octal form here (assuming 2.7 is dropped).

Py2 is not dropped :(.

> manual construction of the octal with a check of 747 or 644 on the basis of extant file executable would read more cleanly here.

Just to be clear this isn't constructing an octal.  This is a zip file spec quirk that the mode is part of the upper 16 bits of `external_attr` so we just need to shift it up to the right place.  I'll add a comment about that.

> note a 755 perserve on executable would not work in lambda, as the user executing the lambda is not the same as the owner of the files afaicr

Not sure what you mean here.  It will work in the sense that you'll be able to execute a 0o755 file on lambda if we preserve that mode.  I confirmed that this is the case.